### PR TITLE
Add http endpoint and prometheus metrics framework

### PR DESCRIPTION
**Pull request by benzcash**
This change adds and http endpoint in addition to the grpc endpoint.

We then register a prometheus handler at `/metrics` path for metric collection.

# Breaking changes

The `-bind-addr` option has been replaced with `-grpc-bind-addr` option. It functions the same, the change is to differentiate it from the new option `-http-bind-addr`.

# Other changes

We need to register the prometheus interceptor in addition to the existing logging interceptor, so the change also adds the chaining middleware, which calls these interceptors in order defined.

Source: https://github.com/grpc-ecosystem/go-grpc-middleware

Check out the other interceptors there for some possible additional functionality!

Once this PR is built, you can verify the functionality by visiting the http://127.0.0.1:9068/metrics url in a browser. On startup, you will only see internal metrics for go and prometheus library. As methods are called on the grpc port, metrics will be generated for those operations.

Basically, the metrics won't show up, until calls are made.

These are just some basic, out of the box metrics. My idea is that once we get this up and running, we can start adding more specific metric collections.

### CODE REVIEW

**Comment by mdr0id** [comment]
I feel like I am missing something trivial, so apologies in advance:

Procedure
1.  fresh clone PR
2. `cp .env.template .env`
3.  `docker-compose up`
     lightwalletd and zcashd containers exit
4. Added `ZCASHD_NETWORK=testnet` to .env to allow zcashd container to not exit
5. Now it appears the node is attempting to sync: `docker logs lightwalletd_zcashd_1` show UpdateTip iterating
6. `docker logs lightwalletd_lightwalletd_1` 
```flag provided but not defined: -grpc-bind-addr```

**Comment by benzcash** [comment]
> docker logs lightwalletd_lightwalletd_1
flag provided but not defined: -grpc-bind-addr

If you have previously built a container for the docker-compose project, it won't rebuild for this branch without being forced.

Try:
```
docker-compose build
docker-compose up -d
```

**Comment by benzcash** [comment]
> NACK if this saves statistics about gRPC calls.
> 
> I ran into some errors trying to test this (pasted below if it's helpful somehow), so I'm not sure exactly what this will log, but from reading https://github.com/grpc-ecosystem/go-grpc-prometheus it seems like this will log and save gRPC statistics.

> ERROR: for grafana  Cannot create container for service grafana: logger: no log driver named 'loki' is registered

These errors are because the loki plugin is not installed.

https://github.com/zcash/lightwalletd/blob/master/docs/docker-compose-setup.md#install-requirements

Specifically, run:
```
docker plugin install grafana/loki-docker-driver:latest --alias loki --grant-all-permissions
```
> Ideally that information should not be saved at all, so that even in case the server is compromised, it won't exist to be stolen by the attacker. We should keep logging (at least for production environments) to the bare minimum required to operate `lightwalletd` and ensure it's working reliably.

The current iteration literally just logs the client message size, how many time each **method** is called (not parameters) and the time to respond. I think that addresses the concern, no?

```
$ curl -s http://127.0.0.1:9068/metrics | grep grpc
# HELP grpc_server_handled_total Total number of RPCs completed on the server, regardless of success or failure.
# TYPE grpc_server_handled_total counter
grpc_server_handled_total{grpc_code="OK",grpc_method="GetLatestBlock",grpc_service="cash.z.wallet.sdk.rpc.CompactTxStreamer",grpc_type="unary"} 1
grpc_server_handled_total{grpc_code="OK",grpc_method="ServerReflectionInfo",grpc_service="grpc.reflection.v1alpha.ServerReflection",grpc_type="bidi_stream"} 1
# HELP grpc_server_handling_seconds Histogram of response latency (seconds) of gRPC that had been application-level handled by the server.
# TYPE grpc_server_handling_seconds histogram
grpc_server_handling_seconds_bucket{grpc_method="GetLatestBlock",grpc_service="cash.z.wallet.sdk.rpc.CompactTxStreamer",grpc_type="unary",le="0.005"} 1
grpc_server_handling_seconds_bucket{grpc_method="GetLatestBlock",grpc_service="cash.z.wallet.sdk.rpc.CompactTxStreamer",grpc_type="unary",le="0.01"} 1
grpc_server_handling_seconds_bucket{grpc_method="GetLatestBlock",grpc_service="cash.z.wallet.sdk.rpc.CompactTxStreamer",grpc_type="unary",le="0.025"} 1
grpc_server_handling_seconds_bucket{grpc_method="GetLatestBlock",grpc_service="cash.z.wallet.sdk.rpc.CompactTxStreamer",grpc_type="unary",le="0.05"} 1
grpc_server_handling_seconds_bucket{grpc_method="GetLatestBlock",grpc_service="cash.z.wallet.sdk.rpc.CompactTxStreamer",grpc_type="unary",le="0.1"} 1
grpc_server_handling_seconds_bucket{grpc_method="GetLatestBlock",grpc_service="cash.z.wallet.sdk.rpc.CompactTxStreamer",grpc_type="unary",le="0.25"} 1
grpc_server_handling_seconds_bucket{grpc_method="GetLatestBlock",grpc_service="cash.z.wallet.sdk.rpc.CompactTxStreamer",grpc_type="unary",le="0.5"} 1
grpc_server_handling_seconds_bucket{grpc_method="GetLatestBlock",grpc_service="cash.z.wallet.sdk.rpc.CompactTxStreamer",grpc_type="unary",le="1"} 1
grpc_server_handling_seconds_bucket{grpc_method="GetLatestBlock",grpc_service="cash.z.wallet.sdk.rpc.CompactTxStreamer",grpc_type="unary",le="2.5"} 1
grpc_server_handling_seconds_bucket{grpc_method="GetLatestBlock",grpc_service="cash.z.wallet.sdk.rpc.CompactTxStreamer",grpc_type="unary",le="5"} 1
grpc_server_handling_seconds_bucket{grpc_method="GetLatestBlock",grpc_service="cash.z.wallet.sdk.rpc.CompactTxStreamer",grpc_type="unary",le="10"} 1
grpc_server_handling_seconds_bucket{grpc_method="GetLatestBlock",grpc_service="cash.z.wallet.sdk.rpc.CompactTxStreamer",grpc_type="unary",le="+Inf"} 1
grpc_server_handling_seconds_sum{grpc_method="GetLatestBlock",grpc_service="cash.z.wallet.sdk.rpc.CompactTxStreamer",grpc_type="unary"} 3.706e-05
grpc_server_handling_seconds_count{grpc_method="GetLatestBlock",grpc_service="cash.z.wallet.sdk.rpc.CompactTxStreamer",grpc_type="unary"} 1
grpc_server_handling_seconds_bucket{grpc_method="ServerReflectionInfo",grpc_service="grpc.reflection.v1alpha.ServerReflection",grpc_type="bidi_stream",le="0.005"} 0
grpc_server_handling_seconds_bucket{grpc_method="ServerReflectionInfo",grpc_service="grpc.reflection.v1alpha.ServerReflection",grpc_type="bidi_stream",le="0.01"} 1
grpc_server_handling_seconds_bucket{grpc_method="ServerReflectionInfo",grpc_service="grpc.reflection.v1alpha.ServerReflection",grpc_type="bidi_stream",le="0.025"} 1
grpc_server_handling_seconds_bucket{grpc_method="ServerReflectionInfo",grpc_service="grpc.reflection.v1alpha.ServerReflection",grpc_type="bidi_stream",le="0.05"} 1
grpc_server_handling_seconds_bucket{grpc_method="ServerReflectionInfo",grpc_service="grpc.reflection.v1alpha.ServerReflection",grpc_type="bidi_stream",le="0.1"} 1
grpc_server_handling_seconds_bucket{grpc_method="ServerReflectionInfo",grpc_service="grpc.reflection.v1alpha.ServerReflection",grpc_type="bidi_stream",le="0.25"} 1
grpc_server_handling_seconds_bucket{grpc_method="ServerReflectionInfo",grpc_service="grpc.reflection.v1alpha.ServerReflection",grpc_type="bidi_stream",le="0.5"} 1
grpc_server_handling_seconds_bucket{grpc_method="ServerReflectionInfo",grpc_service="grpc.reflection.v1alpha.ServerReflection",grpc_type="bidi_stream",le="1"} 1
grpc_server_handling_seconds_bucket{grpc_method="ServerReflectionInfo",grpc_service="grpc.reflection.v1alpha.ServerReflection",grpc_type="bidi_stream",le="2.5"} 1
grpc_server_handling_seconds_bucket{grpc_method="ServerReflectionInfo",grpc_service="grpc.reflection.v1alpha.ServerReflection",grpc_type="bidi_stream",le="5"} 1
grpc_server_handling_seconds_bucket{grpc_method="ServerReflectionInfo",grpc_service="grpc.reflection.v1alpha.ServerReflection",grpc_type="bidi_stream",le="10"} 1
grpc_server_handling_seconds_bucket{grpc_method="ServerReflectionInfo",grpc_service="grpc.reflection.v1alpha.ServerReflection",grpc_type="bidi_stream",le="+Inf"} 1
grpc_server_handling_seconds_sum{grpc_method="ServerReflectionInfo",grpc_service="grpc.reflection.v1alpha.ServerReflection",grpc_type="bidi_stream"} 0.007439815
grpc_server_handling_seconds_count{grpc_method="ServerReflectionInfo",grpc_service="grpc.reflection.v1alpha.ServerReflection",grpc_type="bidi_stream"} 1
# HELP grpc_server_msg_received_total Total number of RPC stream messages received on the server.
# TYPE grpc_server_msg_received_total counter
grpc_server_msg_received_total{grpc_method="GetLatestBlock",grpc_service="cash.z.wallet.sdk.rpc.CompactTxStreamer",grpc_type="unary"} 1
grpc_server_msg_received_total{grpc_method="ServerReflectionInfo",grpc_service="grpc.reflection.v1alpha.ServerReflection",grpc_type="bidi_stream"} 2
# HELP grpc_server_msg_sent_total Total number of gRPC stream messages sent by the server.
# TYPE grpc_server_msg_sent_total counter
grpc_server_msg_sent_total{grpc_method="GetLatestBlock",grpc_service="cash.z.wallet.sdk.rpc.CompactTxStreamer",grpc_type="unary"} 1
grpc_server_msg_sent_total{grpc_method="ServerReflectionInfo",grpc_service="grpc.reflection.v1alpha.ServerReflection",grpc_type="bidi_stream"} 2
# HELP grpc_server_started_total Total number of RPCs started on the server.
# TYPE grpc_server_started_total counter
grpc_server_started_total{grpc_method="GetLatestBlock",grpc_service="cash.z.wallet.sdk.rpc.CompactTxStreamer",grpc_type="unary"} 1
grpc_server_started_total{grpc_method="ServerReflectionInfo",grpc_service="grpc.reflection.v1alpha.ServerReflection",grpc_type="bidi_stream"} 1
```

This data isn't logged on the server in any manner. It only presents the data for a collector to retrieve and store (Prometheus). A compromised server would already make this information available through  tracing.

**Comment by defuse** [comment]
There's two different things I'm worried about:

1. If an attacker can query this http endpoint themselves to retrieve the current count of how many times each method has been called, they can query it repeatedly and to see if there's an increase in calls corresponding to spamming their victim with payments and confirm they know a user's address. The signal might be hard to discern from the noise if there's a large number of active users, but it would still exist. So at the very least the endpoint needs to be authenticated so that only Prometheus can query it.

2. If Prometheus saves the method call counts frequently, then if the Prometheus database is ever compromised, the attacker learns some information about lightwalletd users' activity in the past. It's not obvious how much information leakage there will be. Wallets are currently being written assuming lightwalletd will keep the timing/frequency of method calls private, so it could be nothing now and become worse as we add features before there are significant privacy improvements to the protocol.

A middle ground would be to expose this endpoint, authenticate it so that only Prometheus can query it, and make sure that Prometheus only retains history for a short period (say 1 week) so that the historical data isn't at risk of compromise (this might be difficult, it's hard to delete data such that even root with raw access to the disk can't recover it).

**Comment by defuse** [comment]
Is it possible to whitelist which methods get counted? Certain ones like `GetLatestBlock`, `GetBlock`, and `GetBlockRange` shouldn't have any privacy consequences (beyond possibly revealing what time certain unidentified users were using their wallet, which I think is acceptable).

**Comment by benzcash** [comment]
> A middle ground would be to expose this endpoint, authenticate it so that only Prometheus can query it, and make sure that Prometheus only retains history for a short period (say 1 week) so that the historical data isn't at risk of compromise (this might be difficult, it's hard to delete data such that even root with raw access to the disk can't recover it).

For sure, the plan was to have prometheus use a client certificate and the the node to use the https://github.com/benzcash/prometheus-proxy to validate client request certs.

> Is it possible to whitelist which methods get counted?

Certainly. But the purpose of this monitoring would be determining what calls are getting executed and how long they take. Incomplete information would make the data much less useful, or worse, misleading.

File: cmd/server/main.go [hunk 1]
```diff
@@ -175,7 +176,22 @@ func main() {
...
+			logInterceptor,
+			grpc_prometheus.UnaryServerInterceptor),
+		))
+	grpc_prometheus.EnableHandlingTimeHistogram()
+	grpc_prometheus.Register(server)
+
+	// Start the HTTP server endpoint
+	go func() {
+		startHTTPServer(opts)
+	}()
```
Comment (LarryRuane) [inline cmd/server/main.go]: Trivial simplification, I think you can just: `go startHTTPServer(opts)`

**Comment by defuse** [comment]
The client cert authentication option sounds good to me, as long as we're on the same page that the security assurance level of the systems storing that data need to be on par with how we'd store private user information (if we had any). We probably want @zebambam's input on this since it's more of a company policy kind of decision.

(BTW sorry for the insanely late reply, I'm suffering from notification fatigue and nearly always miss GitHub replies, review requests, etc. until months later. I'm trying to fix that by looking at my GitHub inbox every day).

**Comment by LarryRuane** [comment]
Should any changes be made to `README.md`, especially as relates to deployment, as part of this PR?

File: go.mod [hunk 2]
```diff
@@ -5,20 +5,17 @@ go 1.12
...
+	github.com/grpc-ecosystem/go-grpc-prometheus v1.2.0
 	github.com/pkg/errors v0.8.1
+	github.com/prometheus/client_golang v0.9.3
 	github.com/sirupsen/logrus v1.4.2
-	github.com/smartystreets/assertions v1.0.1 // indirect
 	github.com/spf13/cobra v0.0.6
 	github.com/spf13/viper v1.6.2
-	github.com/stretchr/testify v1.4.0 // indirect
-	golang.org/x/crypto v0.0.0-20191122220453-ac88ee75c92c // indirect
+	github.com/zcash-hackworks/lightwalletd v0.3.0
```
Comment (LarryRuane) [inline go.mod:15]: There shouldn't be any references to `zcash-hackworks` in this file (or anywhere here). If you run `make update-grpc` it should fix these. (This makefile target name is a bit misleading, because it updates more than just grpc stuff.)

**Comment by benzcash** [comment]
Opening a new PR, closing this.
